boltdb/bolt_store.go

```diff
@@ -3,413 +3,169 @@ package boltdb
 import (
 	"bytes"
 	"fmt"
-	"github.com/bmeg/arachne/aql"
-	"github.com/bmeg/arachne/gdbi"
+	"log"
+	//"github.com/bmeg/arachne/aql"
+	//"github.com/bmeg/arachne/gdbi"
+	"github.com/bmeg/arachne/kvgraph"
 	"github.com/boltdb/bolt"
-	proto "github.com/golang/protobuf/proto"
-	//"github.com/golang/protobuf/ptypes/struct"
-	//"log"
 )
 
-//Outgoing edges
-//key: src 0x00 dest 0x00 edgeid
-//value: edge properties
-var OEdgeBucket = []byte("oedges")
+var GRAPH_BUCKET = []byte("graph")
 
-//Incoming edges
-//key: dest 0x00 src 0x00 edgeid
-//value: blank
-var IEdgeBucket = []byte("iedges")
-
-//Vertices
-//key: vertex id
-//value: vertex properties
-var VertexBucket = []byte("vertices")
-
-type BoltArachne struct {
-	db *bolt.DB
-}
-
-func NewBoltArachne(path string) gdbi.ArachneInterface {
+func BoltBuilder(path string) (kvgraph.KVInterface, error) {
+	log.Printf("Starting BOLTDB")
 	db, _ := bolt.Open(path, 0600, nil)
-
 	db.Update(func(tx *bolt.Tx) error {
-		if tx.Bucket(OEdgeBucket) == nil {
-			tx.CreateBucket(OEdgeBucket)
-		}
-		if tx.Bucket(IEdgeBucket) == nil {
-			tx.CreateBucket(IEdgeBucket)
-		}
-		if tx.Bucket(VertexBucket) == nil {
-			tx.CreateBucket(VertexBucket)
+		if tx.Bucket(GRAPH_BUCKET) == nil {
+			tx.CreateBucket(GRAPH_BUCKET)
 		}
 		return nil
 	})
-	return &BoltArachne{
+	return &BoltKV{
 		db: db,
-	}
+	}, nil
 }
 
-func (self *BoltArachne) Query() gdbi.QueryInterface {
-	return gdbi.NewPipeEngine(self, false)
+var Loaded error = kvgraph.AddKVDriver("bolt", BoltBuilder)
+
+type BoltKV struct {
+	db *bolt.DB
 }
 
-func (self *BoltArachne) Close() {
-	self.db.Close()
+func (self *BoltKV) Close() error {
+	return self.db.Close()
 }
 
-func (self *BoltArachne) SetVertex(vertex aql.Vertex) error {
+func (self *BoltKV) Delete(id []byte) error {
 	err := self.db.Update(func(tx *bolt.Tx) error {
-		b := tx.Bucket(VertexBucket)
-		d, _ := proto.Marshal(&vertex)
-		b.Put([]byte(vertex.Gid), d)
-		return nil
+		b := tx.Bucket(GRAPH_BUCKET)
+		return b.Delete(id)
 	})
 	return err
 }
 
-func (self *BoltArachne) GetVertex(key string, loadProp bool) *aql.Vertex {
-	var out *aql.Vertex = nil
-	err := self.db.View(func(tx *bolt.Tx) error {
-		b := tx.Bucket(VertexBucket)
-		o := &aql.Vertex{}
-		d := b.Get([]byte(key))
-		if d == nil {
-			return nil
+func (self *BoltKV) DeletePrefix(id []byte) error {
+	err := self.db.Update(func(tx *bolt.Tx) error {
+		b := tx.Bucket(GRAPH_BUCKET)
+		odel := make([][]byte, 0, 100)
+		c := b.Cursor()
+		for k, _ := c.Seek([]byte(id)); bytes.HasPrefix(k, []byte(id)); k, _ = c.Next() {
+			odel = append(odel, k)
 		}
-		if loadProp {
-			proto.Unmarshal(d, o)
-			out = o
-		} else {
-			out = &aql.Vertex{Gid: key}
+		for _, okey := range odel {
+			b.Delete(okey)
```

Comment:
Did you try putting this in the loop above? Seems like it would work there, be simpler, and you'd allocate less memory.

Reply:
Trying to avoid scanning the index and deleting elements from it at the same time. 
A better version at https://github.com/bmeg/arachne/blob/master/badgerdb/badger_graph.go#L188 deletes blocks of keys at a time. Should refactor this code to match.

Reply:
Right. The transaction the code is in should be handling that for you. Not positive, but worth a shot? Up to you.

Reply:
Some of the keyvalue stores really hate it when you alter an index as an iterator is scanning it, even inside the context of a transaction. If you end up deleting a key before you've iterated to the next one, unknown things may happen.